### PR TITLE
Symbol outline / search: Customizable display text

### LINKED ISSUE #220: No document links in catalog's <system uri="..." />
While testing the examples from https://raw.githubusercontent.com/microclimate-devops/xml-language-server/master/client-vscode/xml-client/readme-resources/examples.zip, when editing catalog/catalog.xml, I don't get an hyperlink from uri="invoice.xsd" to the actual file.

```xml
<?xml version="1.0"?>
    <catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog">
    <system systemId="http://invoice.xsd" uri="invoice.xsd"/>
</catalog> 
```

### PR DESCRIPTION
**Pull request by angelozerr**
Symbol outline / search: Customizable display text

Fixes #220

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
To play with this PR, please use https://github.com/redhat-developer/vscode-xml/pull/344 on vscode-xml side.

See my explanation with filter at https://github.com/redhat-developer/vscode-xml/issues/220#issuecomment-708612353

**Comment by angelozerr** [comment]
Please see documentation in https://github.com/redhat-developer/vscode-xml/pull/346/files#diff-a1a54c198023b2a813561e299446d0cebea6ebafef9dd59f8107aaa05f6ac057

**Review by datho7561 — CHANGES_REQUESTED** [review]
One other thing I noticed is that there is an NPE if an expression is missing its "xpath" element in the JSON. This means that no symbols get shown in this case.

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLSymbolInformationsTest.java [hunk 1]
```diff
@@ -12,400 +12,332 @@
...
+		expressionFilter.setXpath("/foo/bar/text()");
+		testSymbolInformationsFor(xml, testURI, symbolSettings, //
+				si("foo", SymbolKind.Field, l(testURI, r(0, 0, 3, 6)), ""), //
+				si("bar: ABCD", SymbolKind.Field, l(testURI, r(1, 1, 1, 46)), "foo"), //
+				si("baz", SymbolKind.Field, l(testURI, r(2, 1, 2, 54)), "foo"));
 
-	private Location createLocation(String uri, int startOffset, int endOffset, DOMDocument xmlDocument) {
-		Range range = createRange(startOffset, endOffset, xmlDocument);
-		return new Location(uri, range);
 	}
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLSymbolInformationsTest.java:342]: done

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSymbolsProvider.java [hunk 2]
```diff
@@ -169,13 +181,13 @@ public DocumentSymbolsResult findDocumentSymbols(DOMDocument xmlDocument, XMLSym
...
-			CancelChecker cancelChecker) throws BadLocationException {
-		if (!isNodeSymbol(node)) {
+			XMLSymbolFilter filter, CancelChecker cancelChecker) throws BadLocationException {
+		if (!isNodeSymbol(node, filter)) {
 			return;
 		}
 		cancelChecker.checkCanceled();
 
-		boolean hasChildNodes = node.hasChildNodes();
+		boolean hasChildNodes = node.hasChildNodes() || node.hasAttributes();
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSymbolsProvider.java]: You are right, but I needed that to instanicate the symbols list too. My code was not clear, I updated it and I introduce a collectAttributes boolean to manage this case.

**Review by datho7561 — APPROVED** [review]
It is good, but there are just a few things that need to be fixed on the client side before merging.

**Comment by angelozerr** [comment]
Thanks @datho7561 for your great review!

**Pull request merged**